ReactNativeClient/lib/registry.js

```diff
@@ -56,140 +56,152 @@ reg.syncTarget = (syncTargetId = null) => {
 // sure it gets synced. So we wait for the current sync operation to
 // finish (if one is running), then we trigger a sync just after.
 reg.waitForSyncFinishedThenSync = async () => {
-	const synchronizer = await reg.syncTarget().synchronizer();
-	await synchronizer.waitForSyncToFinish();
-	await reg.scheduleSync(0);
+	reg.waitForReSyncCalls_.push(true);
+	try {
+		const synchronizer = await reg.syncTarget().synchronizer();
+		await synchronizer.waitForSyncToFinish();
+		await reg.scheduleSync(0);
+	} finally {
+		reg.waitForReSyncCalls_.pop();
```

Comment:
If we are waiting to shutdown sync, and this method is also waiting (`on waitForSyncToFinish()`) then there could be a problem.

When sync finishes, if this method runs first before shutdown, then no problem.

But if it runs the other way round, then this would be a problem (we've pulled down the sync, but this method is using it).  So when shutting down, we give way so that this method can also complete before proceeding with the shutdown.

Reply:
> unfortunately hard to follow the diff in this critical part of Joplin

Yes, sorry about that.  I needed to wrap the code in try/finally to avoid memory leaks, and the diff is a mess because of that.